static/js/components/MemoryLimitSelector.tsx

```diff
@@ -0,0 +1,120 @@
+import React, { Dispatch, FC, SetStateAction } from "react";
+import {
+  Col,
+  Input,
+  Label,
+  RadioInput,
+  Row,
+  Select,
+} from "@canonical/react-components";
+import { useQuery } from "@tanstack/react-query";
+import { fetchResources } from "../api/server";
+import { queryKeys } from "../util/queryKeys";
+import { NotificationHelper } from "../types/notification";
+import { MEM_UNITS, MemoryLimit, MEM_LIMIT_TYPE } from "../types/limits";
+import { DEFAULT_MEM_LIMIT } from "../util/defaults";
+
+interface Props {
+  notify: NotificationHelper;
+  memoryLimit: MemoryLimit;
+  setMemoryLimit: Dispatch<SetStateAction<MemoryLimit>>;
+}
+
+const MemoryLimitSelector: FC<Props> = ({
+  notify,
+  memoryLimit,
+  setMemoryLimit,
+}) => {
+  const { data: resources, error } = useQuery({
+    queryKey: [queryKeys.resources],
+    queryFn: fetchResources,
+  });
+
+  if (error) {
+    notify.failure("Could not load resources.", error);
+  }
+
+  const maxMemory = resources?.memory.total;
+
+  const getMemUnitOptions = () => {
+    return Object.values(MEM_UNITS).map((unit) => ({
+      label: unit,
+      value: unit,
+    }));
+  };
+
+  return (
+    <>
+      <Label>Memory limit</Label>
+      <RadioInput
+        label="fixed (bytes)"
+        checked={memoryLimit.selectedType === MEM_LIMIT_TYPE.FIXED}
+        onChange={() =>
+          setMemoryLimit({
+            ...DEFAULT_MEM_LIMIT,
```

Comment:
I think the fields in the `MemoryLimit` apart from type should be optional. Then this call is simplified to `setMemoryLimit({ selectedType: MEM_LIMIT_TYPE.FIXED })`. So we don't need to spread the default with null values into it.

This also applies to the CpuLimit radio options.

Reply:
If we don't spread the default null values, the old value will be kept. This will result in a dirty object and possibly wrong values being applied, at least for the `MemoryLimit`. Unless we differentiate explicitly the value field for the 2 cases, setting a `percentageValue` field and a `bytesValue` field.

Reply:
After a `setMemoryLimit({ selectedType: MEM_LIMIT_TYPE.FIXED })` the `memoryLimit` object from upstream will not have a `value` property. I don't see how the old value will be kept?

Reply:
Right, I missed the fact that you were passing a whole object there.

Reply:
But there is a problem with the memory type, because we'd fall back to bits and want GiB to be selected, which comes from the default spread. I think we should make it explicit in this case, so when fixed is selected, we also set the unit to GiB.